New data

wpreimes opened the pull request:
- Fix segfault by splitting the tests...could have to do with parallel processing in pytest and/or ISMN metadata generation.
- Add C3S v201912, CCI v4.7, CCI v5.2 (placeholder) and SMAP v6 datasets

wpreimes commented:
Just need to find out why the old CCI testdata is currently failing here before merging...

wpreimes commented:
Also the link stuff in the test data package was not a good idea might cause segfaults, I just fixed segfault (which dont appear if you call single tests) on my setup by replacing links in testdata by actual files...

wpreimes commented:
I would like to merge this as soon as possible, @awst-baum if you could help with it at some point soon that would be really great. If you think that my fixes here are too hacky (test splitting etc), I can try to improve them.

awst-baum commented:
@wpreimes I'm currently working on an urgent customer request. I hope to have a look at the PR before the week is over, though.

awst-baum commented on .travis.yml:
```diff
@@ -46,7 +46,9 @@ install:
-  - pytest -m "not needs_advanced_setup"
+  - pytest -m "not needs_advanced_setup and not long_running and not graphs"
+  - pytest -m "not needs_advanced_setup and long_running and not graphs"
+  - pytest -m "not needs_advanced_setup and long_running and graphs"
```
I'm not very fluent in the travis syntax but isn't the last line repeating the long_running tests unnecessarily?

wpreimes replied:
hmm yes that should work

awst-baum commented on validator/tests/test_validation.py:
```diff
@@ -204,14 +204,15 @@ def check_results(self, run):
         boxplot_pngs = [ x for x in os.listdir(outdir) if fnmatch.fnmatch(x, 'boxplot*.png')]
         self.__logger.debug(boxplot_pngs)
-        assert len(boxplot_pngs) == 12
+        assert len(boxplot_pngs) == n_metrics
```
Good idea, I like :heart:

awst-baum commented on validator/tests/test_validation.py:
```diff
@@ -568,6 +569,7 @@ def test_setup_filtering_max(self):
         start_time = time.time()
 
         for dataset in Dataset.objects.all():
+            print(dataset.pretty_name)
```
The print statements in the test_setup_filtering_max test should probably be replaced by log statements, a la
`self.__logger.info(dataset.pretty_name)`
At least this makes it much easier to search for test output in valentina.log on my system.

awst-baum commented on validator/validation/graphics.py:
```diff
@@ -124,7 +124,8 @@ def generate_boxplot(validation_run, outfolder, variable, label, values, unit_re
 
-def generate_overview_map(validation_run, outfolder, metric, label, values, dc1, dc2, pair_name, unit_ref, lons, lats):
+def generate_overview_map(validation_run, outfolder, metric, label, values, dc1, dc2,
+                          pair_name, unit_ref, lons, lats, draw_grid=True):
```
What's with the draw_grid parameter? Do I remember correctly that you mentioned that drawing the grid can be problematic for the Hawaiian testcase?

wpreimes replied:
that was for the new plotting image, here I added the switch just because of past experience I thought it could cause problems. Drawing the grid is currently switched off in the tests, but I can try turning it on again.

awst-baum replied:
If the grid doesn't make problems in testing, I'd try to keep the tests as close to the ops reality as possible (i.e. with grid if the graphs in ops are with grid). If it breaks the tests, I'd leave the switch in.

wpreimes replied:
ok, guess that was all part of finding out what caused the segfaults

awst-baum commented on validator/validation/graphics.py:
```diff
@@ -155,10 +156,13 @@ def generate_overview_map(validation_run, outfolder, metric, label, values, dc1,
-        lats_map = safe_arange(extent[3], extent[2], dy)
-        lons_map = safe_arange(extent[0], extent[1], dx)
+            lats_map = np.arange(extent[3], extent[2], dy)
+            lons_map = np.arange(extent[0], extent[1], dx)
```
The duplication of lats_map/lons_map in both branches of the if-else clause looks like a classic tripping hazard for future maintainers of this code. Why safe_arange in one case but np.arange in the other? Ideal place to hide some bugs, is it really necessary?

wpreimes replied:
yes, `np.arange` cause problems with the 0.1 resolution but works fine for 0.25, `safe_arange` always works (might be a bit slower), but I agree we should be consistent.

awst-baum replied:
So revert this change?

wpreimes replied:
yes, not sure why I even changed it...

awst-baum commented on validator/validation/graphics.py:
```diff
@@ -246,42 +253,43 @@ def generate_all_graphs(validation_run, outfolder):
     # get units for plot labels
     unit_ref = validation_run.reference_configuration.dataset.short_name
 
+    ds =  netCDF4.Dataset(validation_run.output_file.path, mode='r')
```
Is there a particular reason why you removed the 'with' keyword and close manually at the end? My gut feeling is that leaving a netcdf file hanging without closing would be a prime candidate for causing a segfault later on. Since you close at the end, it shouldn't make a difference - unless you leave the block somewhere before the end, e.g. when an error occurs.
Could you test if the segfault goes away when you open the netcdf dataset using 'with'?

wpreimes replied:
thats just the way how I usually do it without getting segfaults ;-) The file gets closed also upon a crash, so it should never stay opened... I don't think it makes a difference, therefore I have no problem reverting this if the old way turns out to work as well.

awst-baum replied:
Yeah, I also suspect it doesn't or shouldn't make a difference, I just learned that using 'with' was more pythonic than not doing it.

awst-baum commented:
My hunch is still that the segfault is netcdf-related. If you're bored/desperate, you could try playing around with the various parameters to [Dataset.__init__](https://unidata.github.io/netcdf4-python/netCDF4/index.html#netCDF4.Dataset.__init__) like `diskless`, `keepweakref`, `parallel` and see if that improves things - but it's just a wild guess on my side and I don't completely understand yet how those options _should_ be used.

wpreimes commented:
> I've been trying to reproduce the segfault for a while now, currently it's a classic case of "works for me".
> On my system, pytest hangs after successfully executing the tests (`pytest -m "not needs_advanced_setup"`) but no segfault and no other problems...

Btw, pytests also hangs on my develop setup.

```
kombu.exceptions.OperationalError: [Errno 111] Connection refused
```

awst-baum commented:
The kombu message sounds like you want to do something with celery, e.g. the advanced setup tests.

wpreimes commented:
I think I have changed everything as we said, still no segfault, neither here, nor locally for me. Maybe the issue had to do with the symlinks in the testdata, which was actually the last thing that I changed if I remember correctly.
